architecture-decision-records/0022-cloudfront-upload-proxy.md

```diff
@@ -0,0 +1,164 @@
+# 22. Proxy S3 uploads through CloudFront
+
+**Date:** 2020-09-17
+
+## Context
+
+When a user uploads a file to TDR, the file is currently sent from the JavaScript running in the user's browser straight
+to the AWS S3 upload bucket. This means that there are several HTTP requests from the user's browser to AWS domains:
+
+- `cognito-identity.eu-west-2.amazonaws.com`: Exchange a TDR access token for a Cognito token
+- `sts.amazonaws.com`: Exchange the Cognito token for a long-lived STS token
+- `tdr-upload-files-dirty-<environment>.s3.eu-west-2.amazonaws.com`: Upload the files
+
+The sequence of requests looks like this (simplified slightly - CORS requests and multiple requests to the auth server
+and S3 are omitted for clarity):
+
+![Sequence diagram showing current upload process](images/upload-sequence-diagram-cognito.png)
+
+Some TDR users are on restricted networks, and the upload step completely fails. From what we could see on a video call
+with one user, the network calls to TDR domains are succeeding but the ones to `amazonaws.com` are being blocked.
+
+We think that if we can replace requests to `amazonaws.com` with requests to `tdr.nationalarchives.gov.uk`, these users
+will be able to upload files.
+
+## Options
+
+All the options we considered involved proxying requests to S3 through a service that we could assign a TDR domain name
+to. We didn't consider moving away from S3, because it integrates well with the processes that happen after the file is
+uploaded, such as the file check and export tasks.
+
+### Proxy options
+
+Whatever option we choose, it will have to solve the problem of sending the data to S3, and authorizing the request so
+that users only have access to S3 objects associated with their own consignments. Another constraint is that we'd like
+to keep using the AWS JavaScript SDK, because the [S3 upload] method automatically handles multipart uploads by chunking
+the file and sending the start and end requests to combine the chunks in S3.
+
+We considered running NGINX in ECS, or writing proxy code and running it in Lambda or ECS. We didn't look very far into
+these options because it means maintaining an extra service. We would also have to proxy the requests to Cognito and
+STS, or find some way to use the AWS SDK to send Keycloak tokens with the upload requests.
```

Comment:
It might be worth mentioning that for the Lambda option, we have a problem with the lambda timeouts, so if a request takes more than the 15 minute max then it would fail. I know it's unlikely with the 5Mb chunking but it's a restriction we didn't have before. 
I also looked briefly at API gateway as a front end to S3 but this has a max upload size of 10Mb and a max request timeout of 30 seconds so that was even less useful. Not sure if it's even worth mentioning :smile:

Reply:
Like you say, the chunking means we wouldn't be hitting timeouts. I think we'd have reduced the chunk size if those Lambdas were returning in more than a few seconds.

So would API Gateway work if you chunked the requests? I guess you'd still have the problem of how to authorise the requests - you'd have the same problem on lines 39-40 about proxying extra requests or adding the Keycloak token to AWS JS SDK requests.

Reply:
API Gateway would have worked with the chunked requests, it's the 30s timeout that would have caused issues I think. Over 30s for a 5Mb chunk isn't unreasonable on a slow connection. We wouldn't have had to bother with Cognito or STS as we could have had a custom authoriser but that would have been unlikely to work nicely with the S3 SDK. I'm happy to leave this bit out anyway as it doesn't really add anything.

Reply:
I've added a bit about API Gateway, just in case we need to revisit the decision. Does this make sense?

> We considered using API Gateway as a proxy. As with the Lambda or ECS options, we would either have had to proxy the Cognito and STS requests too, or found a way to add the Keycloak token to the AWS JavaScript SDK upload request. API Gateway also has a maximum request time of 30 seconds, which might not be enough for users uploading large files over slow connections. This might have required us to reduce the default chunk size when splitting the file for upload.